File: mns-reactor/src/main/java/cz/fb/manaus/reactor/price/ProbabilityCalculator.java
package cz.fb.manaus.reactor.price;

import cz.fb.manaus.core.model.MarketPrices;
import cz.fb.manaus.core.model.Side;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;

@Component
public class ProbabilityCalculator {

    @Autowired
    private PriceService priceService;

    public Map<Long, Double> fromFairness(double fairness, Side side, MarketPrices prices) {
        var sidePrices = prices.getHomogeneous(side);
        var runnerPrices = sidePrices.getRunnerPrices();

        var result = new HashMap<Long, Double>();
        for (var runnerPrice : runnerPrices) {
            var bestPrice = runnerPrice.getBestPrice();
            var unfairPrice = bestPrice.get().getPrice();
            var fairPrice = priceService.getFairnessFairPrice(unfairPrice, fairness);
            result.put(runnerPrice.getSelectionId(), 1 / fairPrice);
        }
        return result;
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/price/ProbabilityCalculator.kt
package cz.fb.manaus.reactor.price

import cz.fb.manaus.core.model.RunnerPrices
import cz.fb.manaus.core.model.Side

object ProbabilityCalculator {

    fun fromFairness(fairness: Double, side: Side, prices: List<RunnerPrices>): Map<Long, Double> {
        val sidePrices = prices.map { it.by(side) }

        val result = HashMap<Long, Double>()
        for (runnerPrice in sidePrices) {
            val bestPrice = runnerPrice.bestPrice
            val unfairPrice = bestPrice!!.price
            val fairPrice = Pricing.getFairnessFairPrice(unfairPrice, fairness)
            result[runnerPrice.selectionId] = 1 / fairPrice
        }
        return result
    }

}
